File: crates/blazen-llm/src/types/tool_output.rs
//! Two-channel tool output: what the user/caller sees vs. what the LLM
//! sees on the next turn. See the workspace docs for the design rationale.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::types::message::ContentPart;

/// What a tool returns. Carries two channels:
/// - `data`: the typed value the caller sees programmatically (full structured payload).
/// - `llm_override`: explicit override for what to send to the LLM on the next turn.
///   `None` means each provider applies its default conversion from `data`.
///
/// Generic over `T` (defaults to `Value`) so [`crate::TypedTool`] can carry a
/// concrete `Output` type up to the trait boundary, where it gets type-erased
/// to `ToolOutput<Value>` for `dyn Tool` heterogeneity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "tsify", derive(tsify_next::Tsify))]
pub struct ToolOutput<T = Value> {
    pub data: T,

    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub llm_override: Option<LlmPayload>,
}

impl<T> ToolOutput<T> {
    pub fn new(data: T) -> Self {
        Self {
            data,
            llm_override: None,
        }
    }

    pub fn with_override(data: T, override_payload: LlmPayload) -> Self {
        Self {
            data,
            llm_override: Some(override_payload),
        }
    }
}

/// Ergonomic conversion: `Ok(value.into())` from a tool returns
/// `ToolOutput { data: value, llm_override: None }`.
impl From<Value> for ToolOutput<Value> {
    fn from(data: Value) -> Self {
        Self::new(data)
    }
}

/// What to send to the LLM as the tool result body. Provider-aware.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[cfg_attr(feature = "tsify", derive(tsify_next::Tsify))]
#[cfg_attr(feature = "tsify", tsify(into_wasm_abi, from_wasm_abi))]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum LlmPayload {
    /// Plain text — works on every provider universally.
    Text { text: String },

    /// Structured JSON. Anthropic and Gemini consume natively;
    /// `OpenAI` / Responses stringify at the wire boundary.
    Json { value: Value },

    /// Multimodal content blocks (Anthropic supports natively as
    /// `tool_result.content` blocks; `OpenAI` falls back to text;
    /// Gemini falls back to a JSON object).
    Parts { parts: Vec<ContentPart> },

    /// Provider-specific escape hatch. The named provider gets `value`
    /// inserted verbatim into its tool-result body; every other provider
    /// falls back to the default conversion from `ToolOutput::data`.
    ProviderRaw { provider: ProviderId, value: Value },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "tsify", derive(tsify_next::Tsify))]
#[cfg_attr(feature = "tsify", tsify(into_wasm_abi, from_wasm_abi))]
#[serde(rename_all = "snake_case")]
pub enum ProviderId {
    /// `openai` on the wire (single word, matching the canonical product name).
    #[serde(rename = "openai")]
    OpenAi,
    /// `openai_compat` on the wire (any OpenAI-compatible endpoint).
    #[serde(rename = "openai_compat")]
    OpenAiCompat,
    Azure,
    Anthropic,
    Gemini,
    Responses,
    Fal,
}

File: bindings/ruby/lib/blazen/tool_output.rb
# frozen_string_literal: true

require "json"

module Blazen
  # Structured tool-handler output carrying both the user-visible +data+
  # and an optional +llm_override+ for what the LLM sees on the next turn.
  #
  # Tool handlers can return either:
  #   - a plain value (auto-wrapped as +data+ with no override), or
  #   - this +ToolOutput+ helper which produces the structured shape.
  #
  # Mirrors JsToolOutput (Node) and PyToolOutput (Python).
  #
  # @example Returning a structured ToolOutput with a text override
  #   tool_handler = lambda do |name, args|
  #     items = fetch_items(args)
  #     Blazen::ToolOutput.new(
  #       data: { items: items },
  #       llm_override: Blazen::ToolOutput.text("Found #{items.size} items.")
  #     ).to_json
  #   end
  #
  # @example Returning a structured ToolOutput with multimodal parts
  #   Blazen::ToolOutput.new(
  #     data: { ok: true },
  #     llm_override: Blazen::ToolOutput.parts([
  #       { type: "text", text: "summary line" },
  #       { type: "image", image: { source: { source_type: "url", url: image_url }, media_type: "image/png" } }
  #     ])
  #   ).to_json
  class ToolOutput
    # @param data [Object] the user-visible payload (any JSON-serialisable value)
    # @param llm_override [Hash, nil] optional override hash produced by one of
    #   the +Blazen::ToolOutput.text/json/parts/provider_raw+ class methods
    def initialize(data:, llm_override: nil)
      @data = data
      @llm_override = llm_override
    end

    attr_reader :data, :llm_override

    def to_h
      h = { "data" => @data }
      h["llm_override"] = @llm_override if @llm_override
      h
    end

    def to_json(*args)
      to_h.to_json(*args)
    end

    # Build a plain-text +llm_override+ payload.
    def self.text(text)
      { "kind" => "text", "text" => text }
    end

    # Build a structured-JSON +llm_override+ payload.
    def self.json(value)
      { "kind" => "json", "value" => value }
    end

    # Build a multi-part +llm_override+ payload. Each part is a Hash with a
    # +"type"+ discriminator (matching the core +ContentPart+ serde tag —
    # NOT +kind+!). Variants: +text+, +image+, +file+, +audio+, +video+.
    #
    # Example part hashes:
    #   { "type" => "text", "text" => "..." }
    #   { "type" => "image", "image" => { "source" => { "source_type" => "url", "url" => "..." }, "media_type" => "image/png" } }
    def self.parts(parts)
      { "kind" => "parts", "parts" => parts }
    end

    # Build a provider-specific +llm_override+ payload. +provider+ must be
    # one of +"openai"+, +"openai_compat"+, +"azure"+, +"anthropic"+,
    # +"gemini"+, +"responses"+, +"fal"+.
    def self.provider_raw(provider:, value:)
      { "kind" => "provider_raw", "provider" => provider, "value" => value }
    end
  end
end
